flink-table-store-core/src/main/java/org/apache/flink/table/store/file/mergetree/sst/SstFileReader.java

```diff
@@ -122,17 +132,56 @@ public Factory(
                 FileStorePathFactory pathFactory) {
             this.keyType = keyType;
             this.valueType = valueType;
+            this.fileFormat = fileFormat;
             this.pathFactory = pathFactory;
-            RowType recordType = KeyValue.schema(keyType, valueType);
-            this.readerFactory = fileFormat.createReaderFactory(recordType);
+
+            this.valueProjection = Projection.range(0, valueType.getFieldCount()).toNestedIndexes();
+            updateProjectedTypes();
+        }
+
+        public Factory withValueProjection(int[][] projection) {
+            valueProjection = projection;
+            updateProjectedTypes();
+            return this;
         }
 
         public SstFileReader create(BinaryRowData partition, int bucket) {
             return new SstFileReader(
                     keyType,
-                    valueType,
+                    projectedValueType,
                     readerFactory,
                     pathFactory.createSstPathFactory(partition, bucket));
         }
+
+        private void updateProjectedTypes() {
+            int numKeyFields = keyType.getFieldCount();
+            projectedValueType = (RowType) Projection.of(valueProjection).project(valueType);
+
+            // see KeyValue#schema for the exact schema
+            int[][] projection = new int[numKeyFields + 2 + valueProjection.length][];
```

Comment:
Create a method in `int[][] KeyValue.projection(int[][] keyProjection, int[][] valueProjection)`